### PR TITLE
fix(upgrade): two-way binding and listening for event

### PR DESCRIPTION
**Pull request by FDIM**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

It is not possible to have a two-way binding and listening to an event in downgraded component. Bound value would remain outdated and only event listener expression would be evaluated.

Issue Number: #22734

## What is the new behavior?
Bound value will be updated first when new value is emitted and then expression (if any) will be evaluated.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/upgrade/src/common/downgrade_component_adapter.ts [hunk 1]
```diff
@@ -168,41 +168,42 @@ export class DowngradeComponentAdapter {
...
-        const setter = getter.assign;
-        if (assignExpr && !setter) {
-          throw new Error(`Expression '${expr}' is not assignable!`);
-        }
-        const emitter = this.component[output.prop] as EventEmitter<any>;
-        if (emitter) {
-          emitter.subscribe({
-            next: assignExpr ? (v: any) => setter !(this.scope, v) :
+  private subscribeToOutput(
+      output: PropertyBinding, expr: string|null, isAssignment: boolean = false) {
```
Comment (gkalpak) [inline packages/upgrade/src/common/downgrade_component_adapter.ts]: `expr` shouldn't be `null` at this point. Can you change the type (and also remove the check below)?

**Review by gkalpak — CHANGES_REQUESTED** [review]
The change LGTM :+1:

The added test only tests the (now deprecated) dynamic version of `ngUpgrade`. We mainly need tests for the static version (with both `UpgradeModule` and `downgradeComponent()`); e.g. in [downgrade_component_spec.ts][1] and [downgrade_module_spec.ts][2].
(Also, that particular test you modified is already very hard to follow. It would be nice if you could add new tests specifically focused on the fix 🙏)

[1]: https://github.com/FDIM/angular/blob/97bedcee6925517d8f44283797f7c93fb9a0af40/packages/upgrade/test/static/integration/downgrade_component_spec.ts
[2]: https://github.com/FDIM/angular/blob/97bedcee6925517d8f44283797f7c93fb9a0af40/packages/upgrade/test/static/integration/downgrade_module_spec.ts

File: packages/upgrade/src/common/downgrade_component_adapter.ts (hunk 1 above)
Comment (FDIM) [reply packages/upgrade/src/common/downgrade_component_adapter.ts]: I'll check if that doesn't cause any issues as the if before calling this only tests if the property is there, not that it's not null

**Comment by FDIM** [comment]
Yeah, that test is quite something. Sad part is that it was the only one that failed if setupOutputs method did nothing, hence why I've changed it. I'll try to update the specs you mentioned and cover the code I've changed.

File: packages/upgrade/src/common/downgrade_component_adapter.ts (hunk 1 above)
Comment (gkalpak) [reply packages/upgrade/src/common/downgrade_component_adapter.ts]: It would be very weird for `attrs.someProp` to be set to `null` (as one would have a directive to specifically do that, which doesn't make much sense) and this would only matter is `isAssignment` is true. But you have a point; there is no good reason to unnecessarily break a weird usecase :grin:

Let's leave it as is.

Comment (FDIM) [reply packages/upgrade/src/common/downgrade_component_adapter.ts]: That is true, and very unlikely - I hope you don't mind that I've removed the check.

**Comment by FDIM** [comment]
I've reverted the test in dynamic version, added a new spec in `static/integration/downgrade_component_spec.ts` and hoping that all is good now

**Review by gkalpak — APPROVED** [review]
LGTM, except it would be nice to have a similar test for the [dynamic version][1] as well.

[1]: https://github.com/FDIM/angular/blob/1eab1f31d412fb27ae70a90edd3e1b200fe91354/packages/upgrade/test/dynamic/upgrade_spec.ts

**Comment by gkalpak** [comment]
Also, the commit message is a little confusing in that it makes the bug sound specific to `model`/`model-change` (which is not the case).

**Comment by FDIM** [comment]
I've updated the message a bit and also copied and adapted the test to ensure that dynamic version is also working

**Comment by gkalpak** [comment]
LGTM, thx 👍
(Waiting for Travis to go green (seems to have flaked) and will mark for merge.)

**Comment by mhevery** [comment]
http://test/OCL:189645305:BASE:189645396:1521496278724:3167b3c5
